src/Corporation/Corporation.ts

```diff
@@ -75,23 +76,19 @@ export class Corporation {
     this.seedFunded = params.seedFunded ?? false;
   }
 
-  addFunds(amt: number): void {
+  gainFunds(amt: number, source: FundsSource): void {
     if (!isFinite(amt)) {
-      console.error("Trying to add invalid amount of funds. Report to a developer.");
+      console.error("Trying to add invalid amount of funds. Please report to game developer.");
       return;
     }
+    if ((FundsSourceCapEx as readonly FundsSource[]).includes(source)) {
```

Comment:
Array.includes relies on iterating the entire array (e.g. if the expense is opex, it has to iterate the entire capex array to figure that out). Also the type definitions are requiring a type assertion here to allow array.includes to be used.

One way to address this would be by changing the FundsSources to be enums (add to the Corporation/Enums.ts file), and then use the enum helper to check if the source is in the requested enum. e.g. this line would be

```ts
if (getEnumHelper("FundsSourceCapEx").isMember(source)) {
```

Which instead of iterating an array, becomes a Set.has operation which should be more efficient for checking if an item exists within a list. Moving to an enum would also mean changing each usage of the gainFunds / loseFunds so that the source needs to be referenced using Enum.key, instead of as a raw string at each location. Which would also help in cases where it's not immediately clear to everyone whether that particular item fits in as opex or capex. e.g.

```ts
corp.addFunds(1e9, "private equity");

// would turn into:

corp.addFunds(1e9, FundsSourceCapEx.privateEquity);
```

Which would make it clear to anyone reading it that it's considered to be capexm, without needing to refer to the objects to determine which sources are under which categories. Although moving to an enum helper would also mean some additional upfront memory usage for the two enums and their helpers.

Another option would be just making the capex runtime object into a set directly, and then just having a type defined that includes all possible sources, and only using some of them for the capex set. Opex doesn't even need a runtime object since it is unused:

e.g.

```js
// FundsSource.ts, alternate suggestion without changing to enums
// CapEx sources at the top
export type FundsSource = "product development" |
  "division" |
  "office" |
  "warehouse" |
  "upgrades" |
  "bribery" |
  "public equity" |
  "private equity" |
  "hacknet" |
  "force majeure" |
  // OpEx
  "operating expenses" |
  "operating revenue" |
  "dividends" |
  "tea" |
  "parties" |
  "advert" |
  "materials" |
  "glitch in reality";

// Set of all CapEx sources, for checking at runtime.
// "CapEx" sources only affect valuation through a change in funds, and are not seen as profit/loss for earnings projections.
export const CapExSources = new Set<FundsSource>([
  "product development",
  "division",
  "office",
  "warehouse",
  "upgrades",
  "bribery",
  "public equity",
  "private equity",
  "hacknet",
  "force majeure",
])
``` 

Do either of those sound like good changes to you?

Not super thrilled with the names CapEx and OpEx, since the "Ex" in for expenditures, but it's not just for tracking expenditures but revenue as well. But I don't have a better suggestion so I'm fine with the naming how it is, just a very minor gripe I figured I would mention as well.

Reply:
I agree that CapEx/OpEx is not a very clear phrasing. The main distinction is one-time/repeatable or long-term/short-term, so I can change it to a name that better reflects that.

I'd like to keep the style of `gainFunds()` similar to `Player.gainMoney()`, using a string instead of an Enum. But I also don't like repeating the values in multiple places which could get out of sync.

What do you think of this style with a Set built from a const array? I can't figure out a way to define a type as a key of a const Set without also including the array.